src/util/tmpl/fd_map_giant.c

```diff
@@ -0,0 +1,823 @@
+MAP_(key_copy)( MAP_KEY_T *       kd,
+                MAP_KEY_T const * ks ) {
+  (MAP_KEY_COPY( (kd), (ks) ));
+  return kd;
+}
+
+FD_FN_PURE static inline MAP_(iter_t)
+MAP_(iter_init)( MAP_T const * join ) {
+  if( FD_UNLIKELY( !join ) ) return 0UL; /* Debatable */
+  MAP_(private_t) const * map = MAP_(private_const)( join );
+  ulong ele_rem = map->key_max;
+  for( ; ele_rem; ele_rem-- ) if( !(join[ ele_rem-1UL ].MAP_NEXT >> 63) ) break;
+  return ele_rem;
+}
+
+FD_FN_PURE static inline int
+MAP_(iter_done)( MAP_T const * join,
+                 MAP_(iter_t)  ele_rem ) {
+  (void)join;
+  return !ele_rem;
+}
+
+FD_FN_PURE static inline MAP_(iter_t)
+MAP_(iter_next)( MAP_T const * join,
+                 MAP_(iter_t)  ele_rem ) {
+  for( ; ele_rem; ele_rem-- ) if( !(join[ ele_rem-1UL ].MAP_NEXT >> 63) ) break;
+  return ele_rem;
+}
+
+FD_FN_CONST static inline MAP_T *
+MAP_(iter_ele)( MAP_T *      join,
+                MAP_(iter_t) ele_rem ) {
+  return join + ele_rem - 1UL;
+}
+
+FD_FN_CONST static inline MAP_T const *
+MAP_(iter_ele_const)( MAP_T const * join,
+                      MAP_(iter_t)  ele_rem ) {
+  return join + ele_rem - 1UL;
+}
+
+FD_PROTOTYPES_END
+
+#endif
+
+#if MAP_IMPL_STYLE==1 /* need prototypes */
+
+FD_PROTOTYPES_BEGIN
+
+FD_FN_CONST ulong MAP_(align)    ( void );
+FD_FN_CONST ulong MAP_(footprint)( ulong   key_max );
+void *            MAP_(new)      ( void *  shmem, ulong key_max, ulong seed );
+MAP_T *           MAP_(join)     ( void *  shmap );
+void *            MAP_(leave)    ( MAP_T * join  );
+void *            MAP_(delete)   ( void *  shmap );
+
+MAP_T *
+MAP_(insert)( MAP_T *           join,
+              MAP_KEY_T const * key );
+
+MAP_T *
+MAP_(remove)( MAP_T *           join,
+              MAP_KEY_T const * key );
+
+FD_FN_PURE MAP_T *
+MAP_(query)( MAP_T *           join,
+             MAP_KEY_T const * key,
+             MAP_T *           null );
+
+FD_FN_PURE MAP_T const *
+MAP_(query_const)( MAP_T const *     join,
+                   MAP_KEY_T const * key,
+                   MAP_T const *     null );
+
+FD_FN_PURE int
+MAP_(verify)( MAP_T const * join );
+
+FD_PROTOTYPES_END
+
+#else /* need implementations */
+
+#if MAP_IMPL_STYLE==0 /* local only */
+#define MAP_IMPL_STATIC FD_FN_UNUSED static
+#else
+#define MAP_IMPL_STATIC
+#endif
+
+FD_FN_CONST MAP_IMPL_STATIC ulong
+MAP_(align)( void ) {
+  return fd_ulong_max( 128UL, alignof(MAP_T) );
+}
+
+FD_FN_CONST MAP_IMPL_STATIC ulong
+MAP_(footprint)( ulong key_max ) {
+  ulong align = MAP_(align)();
+
+  /* memory layout is:
+
+       2 ulong | pad | list_cnt ulong | map_private_t | key_max map_t | pad
+       <------ meta_footprint, align multiple ------> 
+       <------------------ footprint, align multiple -------------------->
+
+     Noting that list_cnt is at most key_max, footprint is
+     (conservatively) at most:
+
+       2*sizeof(ulong) + align-1 + key_max*sizeof(ulong) + sizeof(map_private_t) + key_max*sizeof(map_t) + align-1
```

Comment:
key_thresh main point is to make sure maps are created with a small enough value of key_max that various arithmetic operations under the hood are guaranteed to never overflow.  Practically, it is such a large value that using a more conservative bound is simpler and still beyond any amount of memory obtainable on this planet.  It mostly exists to prevent common initialization errors like some non-validated command line input in some startup code that isn't properly bounds checked (e.g. `footprint( some_user_provided - 1 )` from causing crazy things to happen downstream.

The flip side is that if wanting an absolute tight bound for key_thresh, we'd have to do a lot more elaborate calc with a lot more tricky overflow handling logic and it doesn't provide any practical extra benefit.  (I originally wrote it that way and it made me throw up a little bit in the back of my mouth.  So just used simpler equally practical bound.)

Reply:
No change here as per above